bin/varnishd/cache/cache_vcl.c

```diff
@@ -629,17 +632,44 @@ vcl_set_state(struct vcl *vcl, const char *state, struct vsb **msg)
 			i = -1;
 		}
 		else {
+			/* only send the WARM event to directors created before
+			 * the WARM event. This assumes no deletes, see #4142
+			 */
+
+			VTAILQ_INIT(&colddirs);
+			VTAILQ_INIT(&warmdirs);
+
 			Lck_Lock(&vcl_mtx);
+			VTAILQ_SWAP(&vcl->director_list, &colddirs, vcldir, list);
 			vcl->temp = VCL_TEMP_WARM;
 			Lck_Unlock(&vcl_mtx);
```

Comment:
Thank you @AlveElde, you are right that this is brittle.

Because `VTAILQ_REMOVE(head, elm, field)` only touches `head` if `elm` is last in the list, this would only affect the case where the last director added before the swap is removed after the concat, but yes, it is another race. Less relevant, but still a race.

This is not the only place where removing directors is a problem, which is why I suggested a solution in #4142.

Maybe it's best to bring that one in first and then update this PR accordingly.